lightningd/lightningd.c

```diff
@@ -180,47 +280,85 @@ static const char *find_my_path(const tal_t *ctx, const char *argv0)
 			errx(1, "Cannot find %s in $PATH", argv0);
 	}
 
+	/*~ The caller just wants the directory we're in.
+	 *
+	 * Note the magic "take()" macro here: it annotates a pointer as "to
+	 * be taken", and the recipient is expected to take ownership of the
+	 * pointer.
+	 *
+	 * Many CCAN and our own routines support this, but if you hand a take()
+	 * to a non-take routine unfortunately you don't get a compile error.
+	 */
 	return path_dirname(ctx, take(me));
 }
+
+/*~ This returns the PKGLIBEXEC path which is where binaries get installed.
+ * Note the 'TAKES' annotation which is merely documentation that it will
+ * take ownership of 'my_path' if the caller hands take() there.
+ */
 static const char *find_my_pkglibexec_path(const tal_t *ctx,
 					   const char *my_path TAKES)
 {
 	const char *pkglibexecdir;
 	pkglibexecdir = path_join(ctx, my_path, BINTOPKGLIBEXECDIR);
+
+	/*~ Sometimes take() can be more efficient, since the routine can
+	 * manipulate the string in place.  This is the case here. */
 	return path_simplify(ctx, take(pkglibexecdir));
 }
+
 /* Determine the correct daemon dir. */
 static const char *find_daemon_dir(const tal_t *ctx, const char *argv0)
 {
 	const char *my_path = find_my_path(ctx, argv0);
-	if (has_all_daemons(my_path))
+	/* If we're running in-tree, all the subdaemons are with lightningd. */
+	if (has_all_subdaemons(my_path))
 		return my_path;
+
+	/* Otherwise we assume they're in the installed dir. */
 	return find_my_pkglibexec_path(ctx, take(my_path));
 }
 
+/*~ We like to free everything on exit, so valgrind doesn't complain.  In some
```

Comment:
it is not the first time you mention valgrind. What is valgrind?

Reply:
OK, I added "(an awesome runtime memory usage detector for C and C++ programs)" to the first reference.